homeassistant/components/waterfurnace.py

```diff
@@ -0,0 +1,91 @@
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/waterfurnace/
+"""
+from datetime import timedelta
+import logging
+import time
+import threading
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_USERNAME, CONF_PASSWORD
+)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+
+REQUIREMENTS = ["waterfurnace==0.1.0"]
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "waterfurnace"
+UPDATE_TOPIC = DOMAIN + "_update"
+CONF_UNIT = "unit"
+SCAN_INTERVAL = timedelta(seconds=10)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_UNIT): cv.string,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, base_config):
+    """Setup waterfurnace platform."""
+    import waterfurnace.waterfurnace as wf
+    config = base_config.get(DOMAIN)
+
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    unit = config.get(CONF_UNIT)
+
+    wfconn = wf.WaterFurnace(username, password, unit)
+    wfconn.login()
+    hass.data[DOMAIN] = WaterFurnaceData(hass, wfconn)
+    hass.data[DOMAIN].start()
+
+    discovery.load_platform(hass, 'sensor', DOMAIN, {}, config)
+    return True
+
+
+class WaterFurnaceData(threading.Thread):
+    """WaterFurnace Data collector.
+
+    This is implemented as a dedicated thread polling a websocket in a
+    tight loop. The websocket will shut itself from the server side if
+    a packet is not sent at least every 30 seconds. The reading is
+    cheap, the login is less cheap, so keeping this open and polling
+    on a very regular cadence is actually the least io intensive thing
+    to do.
+    """
+
+    def __init__(self, hass, client):
+        """Initialize the data object."""
+        super().__init__()
+        self.hass = hass
+        self.client = client
+        self.unit = client.unit
+        self.data = None
+
+    def run(self):
+        """Thread run loop."""
+        # This does a tight loop in sending read calls to the
+        # websocket. That's a blocking call, which returns pretty
+        # quickly (1 second). It's important that we do this
+        # frequently though, because if we don't call the websocket at
+        # least every 30 seconds the server side closes the
+        # connection.
+        while True:
+            try:
+                self.data = self.client.read()
+                self.hass.helpers.dispatcher.dispatcher_send(UPDATE_TOPIC)
+                time.sleep(SCAN_INTERVAL.seconds)
+            except ConnectionError:
+                self.client.login()
+                _LOGGER.error("Lost our connection to websocket, trying again")
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Error updating waterfurnace data.")
```

Comment:
If the credentials are bad, this will just keep trying. That's not efficient.

Reply:
The login happening inside the except handler will make trigger a second exception which will stop it. It's maybe not the most elegant, but it shouldn't loop forever.

Reply:
You're correct. For some reason I thought the second handler was nested. Can you add an exception handler for the key error that logs an error and breaks the loop?

Reply:
ack